File: apps/windows/TeslaSync.App/feature-views/ChargingSection.Model.cs
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutually-exclusive render branch of the <c>ChargingSection</c> surface — the native union of the
/// states the web component participates in
/// (web/src/features/analytics/components/weekly-digest/ChargingSection.tsx). The web source is a pure
/// presentational child of the Weekly-Digest page (it takes <c>metrics</c> + <c>dailyEnergyData</c> props and
/// performs no fetching), so the branches are a direct function of the input <see cref="ChargingSectionModel"/>
/// — the parent page owns the query lifecycle (the web <c>WeeklyDigestPage</c> renders a skeleton while the
/// drives/charging/alerts queries load and a page-level empty state when the whole week has no data, mounting
/// the sections only once data has resolved). There is therefore no fetch-driven error / stale / offline
/// branch to reproduce inside this surface; every branch maps onto a visible surface and none is ever hidden.
/// </summary>
public enum ChargingSectionState
{
    /// <summary>The Weekly-Digest queries have not resolved yet (the parent is still fetching) — skeleton chrome.</summary>
    Loading,

    /// <summary>Resolved with no charging activity for the week — friendly empty state instead of an all-zero chart.</summary>
    Empty,

    /// <summary>At least one charge session / some energy added (web fall-through) — the chart + stats + week badge.</summary>
    Ready,
}

/// <summary>
/// One projected day bucket of the Daily-Energy-Added bar chart — the native analogue of a single
/// <c>DailyEnergyEntry</c> the web <c>useWeeklyDigest</c> hook bins (<c>{ day, energy }</c>). <see cref="Day"/>
/// is the already-localized weekday label the parent supplies (Mon..Sun); <see cref="Energy"/> is the summed
/// energy added for that day, carried verbatim from the web metric and rendered with the web's "kWh" axis
/// label. Pure data — no WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
public sealed record ChargingSectionDailyEnergy(string Day, double Energy);

/// <summary>
/// The render-time data model the <c>ChargingSection</c> view binds to — the native analogue of the web
/// component's two props (<c>metrics: DigestMetrics</c> narrowed to the charging fields the section reads, plus
/// <c>dailyEnergyData</c>) and the parent's fetch flag. The component is presentational; user-facing labels are
/// resolved from the i18n facade by the projection, not passed in. The numeric fields mirror the web digest
/// metrics verbatim (<c>chargingSessionCount</c>, <c>chargeEnergyAdded</c>, <c>avgChargeRate</c>,
/// <c>chargingCost</c>, <c>prevChargeEnergy</c>) and are rendered with the web's exact formatting and unit
/// suffixes — no display conversion happens here, matching the presentational web source. Pure data — no WinUI
/// types — so the projection is unit-tested without a UI host.
/// </summary>
public sealed record ChargingSectionModel(
    bool Loading,
    long ChargingSessionCount,
    double ChargeEnergyAdded,
    double AvgChargeRate,
    double ChargingCost,
    double PrevChargeEnergy,
    IReadOnlyList<ChargingSectionDailyEnergy> DailyEnergy)
{
    /// <summary>The initial model: the Weekly-Digest fetch is in flight and no charging data has arrived yet.</summary>
    public static ChargingSectionModel Pending { get; } =
        new(true, 0, 0, 0, 0, 0, []);

    /// <summary>A resolved model with no charging activity — the empty state.</summary>
    public static ChargingSectionModel Empty { get; } =
        new(false, 0, 0, 0, 0, 0, []);
}

/// <summary>
/// One projected, render-ready charging mini-stat — the native analogue of a single web <c>MiniStat</c> tile
/// (label + value + decorative icon inside a glass panel). <see cref="Label"/> is the localized caption;
/// <see cref="Value"/> is the pre-formatted value string (already carrying its unit suffix / currency symbol so
/// the view never does number math); <see cref="Glyph"/> is the decorative Segoe Fluent glyph; and
/// <see cref="AutomationName"/> is the spoken "<c>{label}: {value}</c>". Pure data.
/// </summary>
public sealed record ChargingSectionStat(string Label, string Value, string Glyph, string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the section for one input model — the native analogue of what the
/// web <c>ChargingSection</c> renders. Holds the active <see cref="State"/>, the "Charging" <see cref="Title"/>
/// + its decorative <see cref="TitleGlyph"/>, the Daily-Energy chart's <see cref="ChartTitle"/> /
/// <see cref="EnergySeriesLabel"/> / <see cref="DailyEnergy"/> buckets and a spoken <see cref="ChartSummary"/>,
/// the four <see cref="Stats"/>, the week-over-week row (<see cref="WeekOverWeekLabel"/>, the success/warning
/// <see cref="WeekOverWeekStatus"/>, the <see cref="WeekOverWeekText"/> percentage-or-dash and its
/// <see cref="WeekOverWeekAutomationName"/>), the empty + loading copy, and the surface
/// <see cref="AutomationName"/>. Pure data so every branch is asserted headlessly.
/// </summary>
public sealed record ChargingSectionDisplay(
    ChargingSectionState State,
    string Title,
    string TitleGlyph,
    string ChartTitle,
    string EnergySeriesLabel,
    IReadOnlyList<ChargingSectionDailyEnergy> DailyEnergy,
    string ChartSummary,
    IReadOnlyList<ChargingSectionStat> Stats,
    string WeekOverWeekLabel,
    StatusKind WeekOverWeekStatus,
    string WeekOverWeekText,
    string WeekOverWeekAutomationName,
    string EmptyMessage,
    string LoadingLabel,
    string AutomationName)
{
    /// <summary>True when there is at least one day bucket to chart (otherwise the view shows an empty-chart note).</summary>
    public bool HasChart => DailyEnergy.Count > 0;
}

/// <summary>
/// Pure projection from a <see cref="ChargingSectionModel"/> to its <see cref="ChargingSectionDisplay"/> — the
/// native port of web/src/features/analytics/components/weekly-digest/ChargingSection.tsx. The branch precedence
/// mirrors the web data lifecycle the parent page drives (loading → empty → ready); the four mini-stats render
/// through <see cref="NumberFormatting"/> with the web's exact precisions and literal unit suffixes
/// (<c>fmtInt</c> sessions, <c>fmtNumber(_, 1) + " kWh"</c>, <c>fmtNumber(_, 1) + " kW"</c>, and
/// <c>formatCurrency(_, 2)</c> = <c>{symbol}{fmtNumber(_, 2)}</c>); and the week-over-week chip reproduces the
/// web badge exactly — <c>variant = chargeEnergyAdded &gt;= prevChargeEnergy ? success : warning</c> and
/// <c>text = prevChargeEnergy &gt; 0 ? `${fmtNumber(pctChange(...), 1)}%` : '—'</c>. Non-finite inputs are
/// coerced to zero (the web <c>safeNumber</c> guard inside <c>fmtNumber</c>). Every label resolves through the
/// i18n facade using the same keys the web source feeds into <c>t(...)</c>. No WinUI types — unit-tested
/// without a UI host.
/// </summary>
public static class ChargingSectionProjection
{
    /// <summary>Decorative lightning glyph (Segoe Fluent — LightningBolt; web <c>Zap</c>).</summary>
    public const string ZapGlyph = "\uE945";

    /// <summary>Decorative activity glyph (Segoe Fluent — activity line; web <c>Activity</c>).</summary>
    public const string ActivityGlyph = "\uE9D2";

    /// <summary>Decorative money glyph (Segoe Fluent — money; the native cost mapping of the web <c>Fuel</c> icon).</summary>
    public const string CostGlyph = "\uE1D3";

    /// <summary>The em-dash the web badge shows when there is no previous-week baseline to compare against.</summary>
    public const string EmDash = "\u2014";

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade + currency.</summary>
    /// <param name="model">The render-time data model (the web props, narrowed to the charging fields).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <param name="currencySymbol">The active currency symbol for the Total Cost stat (defaults to <c>$</c>).</param>
    public static ChargingSectionDisplay Project(
        ChargingSectionModel model,
        ILocalizer localizer,
        string? currencySymbol = null)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        string currency = string.IsNullOrWhiteSpace(currencySymbol) ? "$" : currencySymbol;

        string title = localizer.GetString("analytics.weeklyDigest.chargingSection", "Charging");
        string chartTitle = localizer.GetString("analytics.weeklyDigest.dailyEnergyAdded", "Daily Energy Added (kWh)");
        string energyLabel = localizer.GetString("analytics.weeklyDigest.energyAdded", "Energy Added");
        string emptyMessage = localizer.GetString("chart.noData", "No data available");
        string loadingLabel = localizer.GetString("common.loading", "Loading");

        IReadOnlyList<ChargingSectionDailyEnergy> daily = NormalizeDaily(model.DailyEnergy);
        ChargingSectionState state = SelectState(model, daily);

        IReadOnlyList<ChargingSectionStat> stats = BuildStats(model, localizer, currency);
        string chartSummary = BuildChartSummary(chartTitle, daily);

        string wowLabel = localizer.GetString("analytics.weeklyDigest.energyVsLastWeek", "Energy vs. Last Week");
        StatusKind wowStatus = Safe(model.ChargeEnergyAdded) >= Safe(model.PrevChargeEnergy)
            ? StatusKind.Success
            : StatusKind.Warning;
        string wowText = Safe(model.PrevChargeEnergy) > 0
            ? NumberFormatting.Format(Safe(PctChange(Safe(model.ChargeEnergyAdded), Safe(model.PrevChargeEnergy))), null, 1) + "%"
            : EmDash;
        string wowAutomation = $"{wowLabel}: {wowText}";

        string automationName = BuildAutomationName(state, title, stats, wowLabel, wowText, emptyMessage, loadingLabel);

        return new ChargingSectionDisplay(
            State: state,
            Title: title,
            TitleGlyph: ZapGlyph,
            ChartTitle: chartTitle,
            EnergySeriesLabel: energyLabel,
            DailyEnergy: daily,
            ChartSummary: chartSummary,
            Stats: stats,
            WeekOverWeekLabel: wowLabel,
            WeekOverWeekStatus: wowStatus,
            WeekOverWeekText: wowText,
            WeekOverWeekAutomationName: wowAutomation,
            EmptyMessage: emptyMessage,
            LoadingLabel: loadingLabel,
            AutomationName: automationName);
    }

    /// <summary>Branch precedence from the web data lifecycle: loading → empty → ready.</summary>
    private static ChargingSectionState SelectState(
        ChargingSectionModel model,
        IReadOnlyList<ChargingSectionDailyEnergy> daily)
    {
        if (model.Loading)
        {
            return ChargingSectionState.Loading;
        }

        // A week with no sessions, no energy added and no charted energy has no charging story to tell —
        // collapse to a friendly empty state rather than rendering an all-zero chart and "0" stats.
        bool hasActivity =
            model.ChargingSessionCount > 0
            || Safe(model.ChargeEnergyAdded) > 0
            || HasPositiveEnergy(daily);

        return hasActivity ? ChargingSectionState.Ready : ChargingSectionState.Empty;
    }

    private static IReadOnlyList<ChargingSectionStat> BuildStats(
        ChargingSectionModel model,
        ILocalizer localizer,
        string currency)
    {
        // Web: fmtInt(metrics.chargingSessionCount)
        var sessions = Stat(
            localizer.GetString("analytics.weeklyDigest.sessions", "Sessions"),
            NumberFormatting.Format(model.ChargingSessionCount, null, 0),
            ZapGlyph);

        // Web: `${fmtNumber(metrics.chargeEnergyAdded, 1)} kWh`
        var totalEnergy = Stat(
            localizer.GetString("analytics.weeklyDigest.totalEnergyAdded", "Total Energy Added"),
            NumberFormatting.Format(Safe(model.ChargeEnergyAdded), null, 1) + " kWh",
            ZapGlyph);

        // Web: `${fmtNumber(metrics.avgChargeRate, 1)} kW`
        var avgRate = Stat(
            localizer.GetString("analytics.weeklyDigest.avgChargeRate", "Avg Charge Rate"),
            NumberFormatting.Format(Safe(model.AvgChargeRate), null, 1) + " kW",
            ActivityGlyph);

        // Web: formatCurrency(metrics.chargingCost, 2) = `${currencySymbol}${fmtNumber(amount, 2)}`
        var totalCost = Stat(
            localizer.GetString("analytics.weeklyDigest.totalCost", "Total Cost"),
            currency + NumberFormatting.Format(Safe(model.ChargingCost), null, 2),
            CostGlyph);

        return [sessions, totalEnergy, avgRate, totalCost];
    }

    private static ChargingSectionStat Stat(string label, string value, string glyph) =>
        new(label, value, glyph, $"{label}: {value}");

    private static List<ChargingSectionDailyEnergy> NormalizeDaily(
        IReadOnlyList<ChargingSectionDailyEnergy>? source)
    {
        if (source is null || source.Count == 0)
        {
            return [];
        }

        var list = new List<ChargingSectionDailyEnergy>(source.Count);
        foreach (var entry in source)
        {
            if (entry is null)
            {
                continue;
            }

            list.Add(new ChargingSectionDailyEnergy(entry.Day ?? string.Empty, Safe(entry.Energy)));
        }

        return list;
    }

    private static string BuildChartSummary(string chartTitle, IReadOnlyList<ChargingSectionDailyEnergy> daily)
    {
        if (daily.Count == 0)
        {
            return chartTitle;
        }

        var parts = new List<string>(daily.Count);
        foreach (var entry in daily)
        {
            parts.Add($"{entry.Day} {NumberFormatting.Format(entry.Energy, null, 1)}");
        }

        return $"{chartTitle}: {string.Join(", ", parts)}";
    }

    private static string BuildAutomationName(
        ChargingSectionState state,
        string title,
        IReadOnlyList<ChargingSectionStat> stats,
        string wowLabel,
        string wowText,
        string emptyMessage,
        string loadingLabel) => state switch
        {
            ChargingSectionState.Loading => loadingLabel,
            ChargingSectionState.Empty => emptyMessage,
            _ => BuildReadyAutomationName(title, stats, wowLabel, wowText),
        };

    private static string BuildReadyAutomationName(
        string title,
        IReadOnlyList<ChargingSectionStat> stats,
        string wowLabel,
        string wowText)
    {
        var builder = new StringBuilder(title);
        foreach (var stat in stats)
        {
            builder.Append(CultureInfo.CurrentCulture, $". {stat.Label} {stat.Value}");
        }

        builder.Append(CultureInfo.CurrentCulture, $". {wowLabel} {wowText}");
        return builder.ToString();
    }

    private static bool HasPositiveEnergy(IReadOnlyList<ChargingSectionDailyEnergy> daily)
    {
        foreach (var entry in daily)
        {
            if (entry.Energy > 0)
            {
                return true;
            }
        }

        return false;
    }

    // Web parity: pctChange(current, previous) — guards a zero denominator before the percentage math. In the
    // badge the caller already gates on previous > 0, but the helper stays faithful to the web for reuse.
    private static double PctChange(double current, double previous)
    {
        if (previous == 0)
        {
            return current > 0 ? 100 : 0;
        }

        return (current - previous) / Math.Abs(previous) * 100;
    }

    // The web safeNumber() guard inside fmtNumber: a non-finite value formats as 0 rather than "NaN"/"∞".
    private static double Safe(double value) => double.IsFinite(value) ? value : 0;
}

/// <summary>
/// PII-safe diagnostics for the <c>ChargingSection</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never a session count, energy figure, charge
/// rate or cost — so a diagnostics line can never leak a user's charging behaviour. Thread-safe.
/// </summary>
public sealed class ChargingSectionDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public ChargingSectionDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=ChargingSection</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={ChargingSectionRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>ChargingSection</c> feature surface — the native mirror of the web component
/// at <c>web/src/features/analytics/components/weekly-digest/ChargingSection.tsx</c>.
/// </summary>
public static class ChargingSectionRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "ChargingSection";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/ChargingSection/ChargingSectionModel.kt
// Pure, framework-free model + projection for the ChargingSection feature view — the native analogue of
// everything the web component derives before it returns JSX
// (web/src/features/analytics/components/weekly-digest/ChargingSection.tsx). No Compose, no Android, no
// HTTP: every declaration here is exercised off-device by the :android:testReleaseUnitTest gate, so the
// composable stays a thin render layer over these pure functions.
//
// The web component is purely presentational — its parent (the Weekly Digest page's `useWeeklyDigest` hook)
// computes the `metrics` summary and the seven-day `dailyEnergyData` series and passes them down. From those
// props the component renders one charging panel: a daily-energy bar chart, a four-tile stat row (session
// count, total energy added, average charge rate, total cost), and a week-over-week energy badge. This file
// owns the parts the web expresses inline: the slice of `DigestMetrics` the panel reads, the seven-day
// series, the `useFormatting` currency contract (`currencySymbol + fmtNumber`), the four formatted stat
// strings, the bar-chart series projection, the `pctChange` week-over-week delta with its success/warning
// tone, and the lifecycle projection onto the shared cache-then-network [UiState] (so the surface renders
// every state the P1/S8 layer can carry). `fmtNumber`/`fmtInt` mirror the web `Intl.NumberFormat`
// half-away-from-zero rounding rather than Java's default banker's rounding.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/ChargingSection — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.chargingsection

import io.teslasync.android.data.UiPhase
import io.teslasync.android.data.UiState
import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.longOrNull
import java.math.RoundingMode
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToLong

/** Em dash shown for an absent week-over-week delta — the web `—` fallback. */
internal const val EM_DASH: String = "\u2014"

/** Default currency symbol when the settings document has none (web `useFormatting` `'$'`). */
internal const val DEFAULT_CURRENCY: String = "$"

/** Energy fraction digits — the web `fmtNumber(metrics.chargeEnergyAdded, 1)` precision. */
internal const val ENERGY_DECIMALS: Int = 1

/** Charge-rate fraction digits — the web `fmtNumber(metrics.avgChargeRate, 1)` precision. */
internal const val RATE_DECIMALS: Int = 1

/** Cost fraction digits — the web `formatCurrency(metrics.chargingCost, 2)` literal precision. */
internal const val COST_DECIMALS: Int = 2

/** Week-over-week percent fraction digits — the web `fmtNumber(pctChange(...), 1)` precision. */
internal const val TREND_DECIMALS: Int = 1

/** Energy unit symbol appended to the total-energy stat — the web literal `kWh`. */
internal const val UNIT_KWH: String = "kWh"

/** Power unit symbol appended to the charge-rate stat — the web literal `kW`. */
internal const val UNIT_KW: String = "kW"

/** Percent scale for the week-over-week delta (web `* 100`). */
private const val PERCENT_FULL: Double = 100.0

/**
 * The slice of the web `DigestMetrics` this charging panel reads (web `types.ts`). The five fields are the
 * exact ones the web `ChargingSection` consumes; the rest of `DigestMetrics` (distance, drives, alerts, …)
 * belongs to the sibling digest sections and is intentionally omitted.
 *
 * @property chargeEnergyAdded total energy added across the week, in kWh (web `metrics.chargeEnergyAdded`).
 * @property prevChargeEnergy the previous week's energy added, in kWh — the week-over-week baseline (web
 *   `metrics.prevChargeEnergy`).
 * @property avgChargeRate average charge rate across the week's sessions, in kW (web `metrics.avgChargeRate`).
 * @property chargingCost total charging cost across the week, in the user's currency (web
 *   `metrics.chargingCost`).
 * @property chargingSessionCount number of charging sessions in the week (web `metrics.chargingSessionCount`).
 */
data class ChargingDigestMetrics(
    val chargeEnergyAdded: Double,
    val prevChargeEnergy: Double,
    val avgChargeRate: Double,
    val chargingCost: Double,
    val chargingSessionCount: Long,
) {
    companion object {
        /** The all-zero metrics — the empty-week outcome (every stat reads zero, the delta reads "—"). */
        val ZERO: ChargingDigestMetrics =
            ChargingDigestMetrics(
                chargeEnergyAdded = 0.0,
                prevChargeEnergy = 0.0,
                avgChargeRate = 0.0,
                chargingCost = 0.0,
                chargingSessionCount = 0L,
            )
    }
}

/**
 * One day's energy-added bar — the native mirror of a web `DailyEnergyEntry` (`{ day: string; energy:
 * number }`). [day] is the localized weekday label (the chart X axis) and [energy] is the kWh added that
 * day (the bar height).
 */
data class DailyEnergyPoint(
    val day: String,
    val energy: Double,
)

/**
 * The decoded charging slice this section renders — the native projection of the two web props
 * (`metrics`, `dailyEnergyData`). Built by a host that already holds the computed weekly digest, or via
 * [ChargingSectionProjection.parse] from a raw digest document.
 */
data class ChargingDigestData(
    val metrics: ChargingDigestMetrics,
    val dailyEnergy: List<DailyEnergyPoint>,
) {
    companion object {
        /** The empty-week value — zeroed metrics and no daily bars (each region shows its empty affordance). */
        val EMPTY: ChargingDigestData = ChargingDigestData(ChargingDigestMetrics.ZERO, emptyList())
    }
}

/**
 * The user's currency display preference this surface needs — the native port of the web `useFormatting`
 * read of the `/settings` document. Only the [currencySymbol] is needed: the total-cost stat formats with
 * the literal 2-digit precision (web `formatCurrency(x, 2)`), so the user's `decimal_precision` does not
 * apply here.
 */
data class ChargingCurrencyPrefs(
    val currencySymbol: String,
) {
    companion object {
        /** The `$` default used before settings load (matches the web default). */
        val DEFAULT: ChargingCurrencyPrefs = ChargingCurrencyPrefs(DEFAULT_CURRENCY)

        private const val KEY_CURRENCY_SYMBOL = "currency_symbol"

        /** Resolves the currency symbol from the raw `/settings` document (web `useFormatting`). */
        fun fromSettings(settings: JsonElement?): ChargingCurrencyPrefs {
            val raw = (settings as? JsonObject)?.get(KEY_CURRENCY_SYMBOL) as? JsonPrimitive
            val symbol = raw?.contentOrNull?.trim()
            return ChargingCurrencyPrefs(if (!symbol.isNullOrEmpty()) symbol else DEFAULT_CURRENCY)
        }
    }
}

/**
 * The four already-formatted stat-row values, in web source order — the native mirror of the web `MiniStat`
 * value props. [sessions] is the grouped session count (web `fmtInt`); [totalEnergyAdded] and
 * [avgChargeRate] are the grouped 1-decimal values with their `kWh`/`kW` unit suffixes; [totalCost] is the
 * currency-formatted cost (web `formatCurrency(_, 2)`).
 */
data class ChargingStatValues(
    val sessions: String,
    val totalEnergyAdded: String,
    val avgChargeRate: String,
    val totalCost: String,
)

/**
 * The bar-chart inputs, preserving the received (chronological) day order — the native analogue of the web
 * `<BarChart data={dailyEnergyData}>` binding. [labels] are the weekday X-axis labels; [values] are the
 * per-day kWh added (each normalized to a finite number so a gap never renders as `NaN`). [isEmpty] mirrors
 * the web series guard so the chart region shows its empty message instead of a blank plot.
 */
data class ChargingBarData(
    val labels: List<String>,
    val values: List<Double?>,
    val isEmpty: Boolean,
)

/**
 * The week-over-week energy badge — the native projection of the web `<Badge>` content. [text] is the
 * signed percent change (web `${fmtNumber(pctChange(...), 1)}%`) or the em dash when there is no prior-week
 * baseline (web `metrics.prevChargeEnergy > 0 ? … : '—'`). [positive] selects the badge tone: the web
 * `success` (this week ≥ last week) versus `warning` (this week down), resolved to a palette accent at the
 * Compose boundary.
 */
data class EnergyTrend(
    val text: String,
    val positive: Boolean,
)

/**
 * Pure projection from the section's inputs to its render state — a 1:1 port of the web component's inline
 * derivations and value formatting. Stateless and side-effect-free so it is fully covered by the off-device
 * unit gate; the composable only resolves localized strings, glyphs, accents, and colors and draws what
 * these return.
 */
object ChargingSectionProjection {
    /**
     * Maps the section's `(data, isLoading)` props onto the shared cache-then-network [UiState] (P1/S8).
     * The web component itself has no loading/error surface (its parent owns those); this adapter adds the
     * lifecycle states the host's feed can carry while preserving web precedence:
     *  - loading → [UiPhase.Loading];
     *  - not loading + data present → [UiPhase.Content] (the panel renders the chart, stats, and badge);
     *  - not loading + no data → [UiPhase.Empty] (the panel still renders, with zeroed stats, an empty
     *    chart, and a "—" delta — never a blank box).
     */
    fun projectUiState(
        data: ChargingDigestData?,
        isLoading: Boolean,
    ): UiState<ChargingDigestData> =
        when {
            isLoading -> UiState.loading()
            data != null -> UiState(phase = UiPhase.Content, data = data)
            else -> UiState(phase = UiPhase.Empty)
        }

    /**
     * The four formatted stat-row values — the native mirror of the web `MiniStat` value expressions:
     * `fmtInt(chargingSessionCount)`, `${fmtNumber(chargeEnergyAdded, 1)} kWh`,
     * `${fmtNumber(avgChargeRate, 1)} kW`, and `formatCurrency(chargingCost, 2)`. [locale] drives the
     * grouping/decimal separators; [currency] supplies the cost symbol.
     */
    fun statValues(
        metrics: ChargingDigestMetrics,
        currency: ChargingCurrencyPrefs,
        locale: Locale,
    ): ChargingStatValues =
        ChargingStatValues(
            sessions = formatInt(maxOf(metrics.chargingSessionCount, 0L), locale),
            totalEnergyAdded = "${fmtNumber(metrics.chargeEnergyAdded, ENERGY_DECIMALS, locale)} $UNIT_KWH",
            avgChargeRate = "${fmtNumber(metrics.avgChargeRate, RATE_DECIMALS, locale)} $UNIT_KW",
            totalCost = formatCurrency(metrics.chargingCost, currency.currencySymbol, COST_DECIMALS, locale),
        )

    /**
     * The bar-chart inputs in received (chronological) order — the native analogue of the web `<BarChart
     * data={dailyEnergyData}>` binding. Each energy sample is normalized via [safe] so a non-finite value
     * never plots as `NaN`. [ChargingBarData.isEmpty] is the web series guard (no days → empty message).
     */
    fun barData(dailyEnergy: List<DailyEnergyPoint>): ChargingBarData =
        ChargingBarData(
            labels = dailyEnergy.map { it.day },
            values = dailyEnergy.map { safe(it.energy) },
            isEmpty = dailyEnergy.isEmpty(),
        )

    /**
     * The week-over-week energy badge — a verbatim port of the web `<Badge>` content and variant. The tone
     * is `success` when this week's energy is at least last week's, else `warning` (web
     * `chargeEnergyAdded >= prevChargeEnergy ? 'success' : 'warning'`); the text is the signed percent
     * change, or the em dash when there is no prior-week baseline (web `prevChargeEnergy > 0 ? … : '—'`).
     */
    fun energyTrend(
        metrics: ChargingDigestMetrics,
        locale: Locale,
    ): EnergyTrend {
        val current = safe(metrics.chargeEnergyAdded)
        val previous = safe(metrics.prevChargeEnergy)
        val text =
            if (previous > 0.0) {
                "${fmtNumber(pctChange(current, previous), TREND_DECIMALS, locale)}%"
            } else {
                EM_DASH
            }
        return EnergyTrend(text = text, positive = current >= previous)
    }

    /**
     * Percent change of [current] over [previous] — a verbatim port of the web `pctChange` helper
     * (weekly-digest/helpers.ts): a zero baseline yields 100 when there is any current value, else 0;
     * otherwise `((current - previous) / |previous|) * 100`.
     */
    fun pctChange(
        current: Double,
        previous: Double,
    ): Double {
        if (previous == 0.0) return if (current > 0.0) PERCENT_FULL else 0.0
        return (current - previous) / abs(previous) * PERCENT_FULL
    }

    /**
     * Locale-aware fixed-precision formatting — the native mirror of the web `fmtNumber(value, decimals)`
     * (`Intl.NumberFormat` with equal min/max fraction digits). Groups thousands and rounds half away from
     * zero so the output matches ECMAScript's `halfExpand` rather than Java's default banker's rounding. A
     * non-finite value is coerced to 0 (web `safeNumber`).
     */
    fun fmtNumber(
        value: Double,
        decimals: Int,
        locale: Locale,
    ): String {
        val safeDecimals = decimals.coerceAtLeast(0)
        val pattern = if (safeDecimals > 0) "#,##0." + "0".repeat(safeDecimals) else "#,##0"
        return DecimalFormat(pattern, DecimalFormatSymbols(locale))
            .apply { roundingMode = RoundingMode.HALF_UP }
            .format(safe(value))
    }

    /**
     * Locale-aware grouped integer formatting — the native mirror of the web `fmtInt` (`fmtNumber(v, 0)`).
     * Delegates to [fmtNumber] at 0 decimals so the half-away-from-zero rounding and grouping match the web.
     */
    fun formatInt(
        value: Long,
        locale: Locale,
    ): String = fmtNumber(value + 0.0, 0, locale)

    /**
     * Currency formatting — the web `useFormatting` `currencySymbol + fmtNumber(amount, decimals)` contract.
     * A blank symbol falls back to `$`; a non-finite amount is normalized to 0 (web `safeNumber`).
     */
    fun formatCurrency(
        amount: Double,
        symbol: String,
        decimals: Int,
        locale: Locale,
    ): String = "${symbol.ifBlank { DEFAULT_CURRENCY }}${fmtNumber(amount, decimals, locale)}"

    /**
     * Decodes a raw weekly-digest document into the typed [ChargingDigestData] this section reads — the
     * native port of the web prop wiring (`metrics`, `dailyEnergyData`). The metrics may sit under a
     * `metrics` object or at the document root; either way any missing/malformed branch degrades to a zero /
     * empty value so the section never throws on a partial payload, and day order is preserved.
     */
    fun parse(digestDoc: JsonElement?): ChargingDigestData {
        val root = digestDoc as? JsonObject ?: return ChargingDigestData.EMPTY
        val metricsObj = root.obj(KEY_METRICS) ?: root
        val metrics =
            ChargingDigestMetrics(
                chargeEnergyAdded = metricsObj.double(KEY_CHARGE_ENERGY_ADDED),
                prevChargeEnergy = metricsObj.double(KEY_PREV_CHARGE_ENERGY),
                avgChargeRate = metricsObj.double(KEY_AVG_CHARGE_RATE),
                chargingCost = metricsObj.double(KEY_CHARGING_COST),
                chargingSessionCount = metricsObj.long(KEY_CHARGING_SESSION_COUNT),
            )
        val daily =
            root.array(KEY_DAILY_ENERGY).mapNotNull { element ->
                val obj = element as? JsonObject ?: return@mapNotNull null
                DailyEnergyPoint(day = obj.string(KEY_DAY), energy = obj.double(KEY_ENERGY))
            }
        return ChargingDigestData(metrics = metrics, dailyEnergy = daily)
    }

    /** Web `safeNumber(v)`: the value when it is a finite number, otherwise 0 — so a chart never plots `NaN`. */
    private fun safe(value: Double): Double = if (value.isFinite()) value else 0.0

    private fun JsonObject.obj(key: String): JsonObject? = this[key] as? JsonObject

    private fun JsonObject.array(key: String): JsonArray = this[key] as? JsonArray ?: JsonArray(emptyList())

    private fun JsonObject.string(key: String): String = (this[key] as? JsonPrimitive)?.contentOrNull ?: ""

    private fun JsonObject.double(key: String): Double = (this[key] as? JsonPrimitive)?.doubleOrNull ?: 0.0

    private fun JsonObject.long(key: String): Long {
        val primitive = this[key] as? JsonPrimitive ?: return 0L
        return primitive.longOrNull ?: primitive.doubleOrNull?.roundToLong() ?: 0L
    }

    private const val KEY_METRICS = "metrics"
    private const val KEY_DAILY_ENERGY = "dailyEnergyData"
    private const val KEY_CHARGE_ENERGY_ADDED = "chargeEnergyAdded"
    private const val KEY_PREV_CHARGE_ENERGY = "prevChargeEnergy"
    private const val KEY_AVG_CHARGE_RATE = "avgChargeRate"
    private const val KEY_CHARGING_COST = "chargingCost"
    private const val KEY_CHARGING_SESSION_COUNT = "chargingSessionCount"
    private const val KEY_DAY = "day"
    private const val KEY_ENERGY = "energy"
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the
 * energy, cost, rate, or session count — so a diagnostics line can never leak the fleet's charging habits
 * or spend.
 */
object ChargingSectionDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "ChargingSection"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
